brvs-core/src/main/java/iroha/validation/transactions/plugin/impl/sora/XorWithdrawalLimitReactionPluggableLogic.java

```diff
@@ -0,0 +1,251 @@
+          "Withdrawal account ID must not be neither null nor empty"
+      );
+    }
+
+    this.queryAPI = queryAPI;
+    this.limitHolderAccount = limitHolderAccount;
+    this.limitSetterAccount = limitSetterAccount;
+    this.xorWithdrawalLimitRemainder = xorWithdrawalLimitRemainder;
+    this.withdrawalAccountId = withdrawalAccountId;
+
+    final long timestampDue = getTimestampFrom(
+        irohaQueryHelper,
+        this.limitSetterAccount
+    );
+    final long brvsLastTimestampDue = getTimestampFrom(
+        irohaQueryHelper,
+        this.queryAPI.getAccountId()
+    );
+    final BigDecimal brvsWithdrawalsSum = getLimitFrom(
+        irohaQueryHelper,
+        this.queryAPI.getAccountId()
+    );
+
+    final BigDecimal amountRemaining = getLimitFrom(
+        irohaQueryHelper,
+        this.limitSetterAccount
+    ).subtract(timestampDue == brvsLastTimestampDue ? brvsWithdrawalsSum : BigDecimal.ZERO);
+
+    updateWithdrawalLimits(
+        new XorWithdrawalLimitRemainder(
+            amountRemaining,
+            timestampDue
+        )
+    );
+  }
+
+  private long getTimestampFrom(IrohaQueryHelper irohaQueryHelper, String setterAccountId) {
+    return Long.parseLong(
+        getDetailedValueFrom(
+            irohaQueryHelper,
+            setterAccountId,
+            limitTimeKey
+        )
+    );
+  }
+
+  private BigDecimal getLimitFrom(IrohaQueryHelper irohaQueryHelper, String setterAccountId) {
+    return new BigDecimal(
+        getDetailedValueFrom(
+            irohaQueryHelper,
+            setterAccountId,
+            limitAmountKey
+        )
+    );
+  }
+
+  private String getDetailedValueFrom(IrohaQueryHelper irohaQueryHelper,
+      String setterAccountId,
+      String key) {
+    return irohaQueryHelper.getAccountDetails(
+        this.limitHolderAccount,
+        setterAccountId,
+        key
+    ).get().orElse("0");
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public Pair<Map<String, String>, BigDecimal> filterAndTransform(
+      Iterable<Transaction> sourceObjects) {
+    final Map<String, String> newDetails = StreamSupport
+        .stream(sourceObjects.spliterator(), false)
+        .map(Transaction::getPayload)
+        .map(Payload::getReducedPayload)
+        .filter(it -> limitSetterAccount.equals(it.getCreatorAccountId()))
+        .map(ReducedPayload::getCommandsList)
+        .flatMap(Collection::stream)
+        .filter(Command::hasSetAccountDetail)
+        .map(Command::getSetAccountDetail)
+        .filter(command -> limitHolderAccount.equals(command.getAccountId()))
+        .collect(
+            Collectors.toMap(
+                SetAccountDetail::getKey,
+                SetAccountDetail::getValue
+            )
+        );
+
+    // to care only about new transactions if there is an external update
+    final long lastUpdateTime =
+        newDetails.containsKey(limitTimeKey) ? Long.parseLong(newDetails.get(limitTimeKey)) : 0L;
```

Comment:
`containsKey` and `get` both lookup in map. So, you do it twice
```suggestion
        Optional.ofNullable(newDetails.get(limitTimeKey)).map(Long::parseLong).orElse(0L);
```